test/examples/active_message.c

```diff
@@ -11,304 +11,258 @@
-	if (UCS_OK != status) {
-		fprintf(stderr, "Failed to get interface address.\n");fflush(stderr);
-		goto out_free;
-	}	 
-	
-	if (iface_attr.cap.flags & UCT_IFACE_FLAG_CONNECT_TO_EP) {
-		/* Create new endpoint */
-		status = uct_ep_create(iface, &ep);	
-		if (UCS_OK != status) {
-			fprintf(stderr, "Failed to create endpoint.\n");fflush(stderr);
-			goto out_free;
-		}	 	
-		/* Get endpoint address */
-		status = uct_ep_get_address(ep, ep_addr);	
-		if (UCS_OK != status) {
-			fprintf(stderr, "Failed to get endpoint address.\n");fflush(stderr);
-			goto out_free_ep;
-		}	 	
-	}
-
-	/* Communicate interface and endpoint addresses to corresponding process */
-	MPI_Send(iface_addr, iface_attr.iface_addr_len, MPI_BYTE, partner, 0, MPI_COMM_WORLD);
-	MPI_Recv(iface_addr, iface_attr.iface_addr_len, MPI_BYTE, partner, 0, MPI_COMM_WORLD, &mpi_status);
-	MPI_Send(ep_addr, iface_attr.ep_addr_len, MPI_BYTE, partner, 0, MPI_COMM_WORLD);
-	MPI_Recv(ep_addr, iface_attr.ep_addr_len, MPI_BYTE, partner, 0, MPI_COMM_WORLD, &mpi_status);
-
-	if (iface_attr.cap.flags & UCT_IFACE_FLAG_CONNECT_TO_EP) {
-		/* Connect endpoint to a remote endpoint */
-		status = uct_ep_connect_to_ep(ep, ep_addr);
-	} else if (iface_attr.cap.flags & UCT_IFACE_FLAG_CONNECT_TO_IFACE) {
-		/* Create an endpoint which is connected to a remote interface */
-		status = uct_ep_create_connected(iface, iface_addr, &ep);
-	} else status = UCS_ERR_UNSUPPORTED;
-	if (UCS_OK != status) {
-		fprintf(stderr, "Failed to connect endpoint\n");fflush(stderr);
-		goto out_free_ep;
-	}
-
-	uint8_t id = 0; /* Tag for active message */
-	/*Set active message handler */
-	status = uct_iface_set_am_handler(iface, id, hello_world, arg, UCT_AM_CB_FLAG_DEFAULT);
-	if (UCS_OK != status) {
-		fprintf(stderr, "Failed to set callback.\n");fflush(stderr);
-		goto out_free_ep;
-	}	 	
-	
-	if (0 == rank) {
-		uint64_t header;
-		char payload[8];
-		unsigned length = sizeof(payload);
-		/* Send active message to remote endpoint */
-		status = uct_ep_am_short(ep, id, header, payload, length);  		
-	} else if (1 == rank) {
-		while (holder) { 
-			/* Explicitly progress any outstanding active message requests */
-			uct_worker_progress(worker);
-		}
-	}
+    /* MPI is initially used to swap the endpoint and interface addresses so each
+     * process has knowledge of the others. */
+    int partner;
+    int size, rank;
+    struct sockaddr *ep_addrs;    /* Endpoint addresses (own and peer's) */
+    struct sockaddr *iface_addrs; /* Interface addresses (own and peer's) */
+    ucs_status_t status;          /* status codes for UCS */
+    uct_ep_h ep;                  /* Remote endpoint */
+    ucs_async_context_t async;    /* Async event context manages times and fd notifications */
+    uint8_t id = 0;
+    void *arg;
+    char *tl_name;
+    struct iface_info if_info;
+    int exit_fail = 1;
+
+    if (argc > 1) {
+        tl_name = argv[1];
+    } else {
+        tl_name = "rc";
```

Comment:
i think there should not be a default value - just exit with error if tl was not specified